### PR TITLE
[#187364732] Reset Password via email

### PR DESCRIPTION
**Pull request by Impano1**
**What does this PR do?**
This PR implements the password reset functionality for the application.
**Description of Task to be completed?**
- Implement password reset feature where users can request a password reset if they forget their login credentials.
- Send a reset link to the user's email.
-After validating the token user tries to insert new password.
- Document the feature using Swagger.

 **How should this be manually tested?**
You can test by watching this video
[click here](https://www.loom.com/share/c0ca90aa51c34d91bc67a8fbbaa13038?sid=dbfb6d77-c32f-4cf5-919f-a8632f2820d8)
**Any background context you want to provide?**
The application currently lacks password reset functionality, which is crucial for users who forget their login credentials. This feature will improve user experience by allowing users to regain access to their accounts without manual intervention from support.

**What are the relevant pivotal tracker/Trello stories?**
[Delivers #187364732]

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "start": "tsc && node dist/index.js",
     "dev": "nodemon -L --exec ts-node index.ts --watch src --ext ts,js",
     "build": "tsc && npm run migrate",
-    "test": "npm run migrate && nyc mocha --require ts-node/register test/*.ts --timeout 80000 --exit",
+    "test": "nyc mocha --require ts-node/register test/*.ts --timeout 80000 --exit",
```
Comment (munezerobagira) [inline package.json]: why did you change this line?

File: src/controllers/UserController.ts [hunk 2]
```diff
@@ -215,4 +215,69 @@ export default class UserController {
...
+
+      if (!user) {
+        return res.status(404).send('User not found')
+      }
+
+      const token = generateToken({
+        id: user.id,
+        email: user.email,
+        username: user.username,
+      })
```
Comment (munezerobagira) [inline src/controllers/UserController.ts]: can you check if this token, can't be used use to login

Comment (munezerobagira) [reply src/controllers/UserController.ts]: an attacker can request reset token and use it to login in victim account account, or if possible you can use OTP that is tied to the user

**Review by munezerobagira — CHANGES_REQUESTED** [review]

File: src/controllers/UserController.ts (hunk 2 above)
Comment (munezerobagira) [reply src/controllers/UserController.ts]: or use can use a different secret

**Review by AngeloChristian1 — APPROVED** [review]

File: src/controllers/UserController.ts (hunk 2 above)
Comment (EricNiyo73) [reply src/controllers/UserController.ts]: let me use different secret

**Review by munezerobagira — APPROVED** [review]

**Review by Yvan-David — APPROVED** [review]
good job
